## What is Fake News
In order to understand how we are going to classify news as fake, we must understand what fake news is. So you might ask yourself... what is Fake news?
Fake News is A type of yellow journalism, fake news encapsulates pieces of news that may be hoaxes and is generally spread through social media and other 
online media. This is often done to further or impose certain ideas and is often achieved with political agendas. Such news items may contain false and/or 
exaggerated claims, and may end up being viralized by algorithms, and users may end up in a filter bubble.

In [1]:
#importing the necessary libraries
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [2]:
#now lets read our data into a dataframe
news = pd.read_csv('news.csv')
#we see the number of records
news.shape
#view our first part of our data using the (.head())
news.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [4]:
#now lets get the labels
labels = news.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [5]:
#now what we are gonna do is split our data into testing and training sets
X_train,X_test,y_train,y_test = train_test_split(news['text'],labels,test_size=0.2,random_state=7)

In [6]:
#initializing the TfidfVectorizer with stop words from the English Language and a maximum document frequency of 0.7
vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

In [8]:
#Fit and transfom the train and test set
Vector_train=vectorizer.fit_transform(X_train) 
Vector_test=vectorizer.transform(X_test)

Next, we’ll initialize a PassiveAggressiveClassifier. This is. We’ll fit this on Vector_train and y_train.

Then, we’ll predict on the test set from the TfidfVectorizer and calculate the accuracy with accuracy_score() from sklearn.metrics.

In [11]:
#initialize Passive Agressive Classifier
pac =PassiveAggressiveClassifier(max_iter=50)
pac.fit(Vector_train,y_train)

#predict on the test Set and calculate accuracy
y_pred = pac.predict(Vector_test)
score = accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 93.13%


We got an accuracy of 93.13% with this model. Finally, let’s print out a confusion matrix to gain insight into the number of false and true negatives and positives.

In [12]:
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[592,  46],
       [ 41, 588]], dtype=int64)

So with this model, we have 592 true positives, 588 true negatives, 41 false positives, and 49 false negatives.